doc/services/storage/zms/zms.rst

```diff
@@ -0,0 +1,432 @@
+
+``GC_done ATE:`` is written to indicate that the next sector has been already garbage
+collected. This ATE could be in any position of the sector.
+
+``ID-ATE:`` are entries that contain a 32 bits Key and describe where the data is stored, its
+size and its crc32
+
+``Data:`` is the actual value associated to the ID-ATE
+
+How does ZMS work?
+******************
+
+Mounting the Storage system
+===========================
+
+Mounting the storage starts by getting the flash parameters, checking that the file system
+properties are correct (sector_size, sector_count ...) then calling the zms_init function to
+make the storage ready.
+
+To mount the NVS filesystem some elements in the zms_fs structure must be initialized.
+
+.. code-block:: c
+
+	struct zms_fs {
+		/** File system offset in flash **/
+		off_t offset;
+
+		/** Storage system is split into sectors, each sector size must be multiple of
+		 * erase-blocks if the device has erase capabilities
+		 */
+		uint32_t sector_size;
+		/** Number of sectors in the file system */
+		uint32_t sector_count;
+
+		/** Flash device runtime structure */
+		const struct device *flash_device;
+	};
+
+Initialization
+==============
+
+As ZMS has a fast-forward write mechanism, we must find the last sector and the last pointer of
+the entry where it stopped the last time.
+It must look for a closed sector followed by an open one, then within the open sector, it finds
+(recover) the last written ATE (Allocation Table Entry).
+After that, it checks that the sector after this one is empty, or it will erase it.
+
+ZMS ID-Data write
+===================
+
+To avoid rewriting the same data with the same ID again, it must look in all the sectors if the
+same ID exist then compares its data, if the data is identical no write is performed.
+If we must perform a write, then an ATE and Data (if not a delete) are written in the sector.
+If the sector is full (cannot hold the current data + ATE) we have to move to the next sector,
+garbage collect the sector after the newly opened one then erase it.
+Data size that is smaller or equal to 8 bytes are written within the ATE.
+
+ZMS ID/data read (with history)
+===============================
+
+By default it looks for the last data with the same ID by browsing through all stored ATEs from
+the most recent ones to the oldest ones. If it finds a valid ATE with a matching ID it retrieves
+its data and returns the number of bytes that were read.
+If history count is provided that is different than 0, older data with same ID is retrieved.
+
+ZMS free space calculation
+==========================
+
+ZMS can also return the free space remaining in the partition.
+However, this operation is very time consuming and needs to browse all valid ATEs in all sectors
+of the partition and for each valid ATE try to find if an older one exist.
+We do not recommend applications to use this function very often at runtime as it could slow
+very much the calling thread
```

Comment:
```suggestion
It is not recommended for application to use this function often, as it is time consuming and could slow down the calling thread.
```
Just a suggestion. Trying to find a better alternative to "could slow very much the calling thread" which is not using a common English word-order...